sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/aggregator/DoubleSumAggregator.java

```diff
@@ -55,6 +55,15 @@ public DoubleAccumulation accumulateDoubleMeasurement(
     return DoubleAccumulation.create(value);
   }
 
+  @Override
+  public void validateAsyncAccumulation(
+      DoubleAccumulation previousAccumulation, DoubleAccumulation currentAccumulation)
+      throws Exception {
+    if (isMonotonic() && currentAccumulation.getValue() < previousAccumulation.getValue()) {
+      throw new Exception("Values must be monotonically increasing.");
```

Comment:
[blocking] Very much against throwing here for various reasons.


1. It is *expected* that counters are allowed to roll-over (i.e. go from Int.Max to Int.Min).  Backends detect this reset.   At a minimum we can detect this issue and reset a start timestamp when it happens.   
2. We *should not prevent this use case*.  For high-rate counters (e.g. request count during high load) this will lead to exceptions in async insturments.

Reply:
> At a minimum we can detect this issue and reset a start timestamp when it happens.

I don't believe we have any precedent for reseting the start timestamp yet - I'm interested in ideas on how this would work. 

> this will lead to exceptions in async insturments.

It's only throwing because throwing and catching is the idiomatic way to indicate an error has occurred. In reality, the only caller of this (AsynchronousMetricStorage) catches and logs, and skips recording the value when this occurs. 

We talked in this in the 12/16 6pm PST SIG. The SDK is in a tough spot with async counters because there's no natural way to validate that values are monotonically increasing. This approach checks in the aggregator, which means that the validation rules change based on how you're aggregating, which isn't right - an async counter doesn't all the sudden get to break monotonicity because the aggregation is last value. And to your point, roll-overs are valid - how do we differentiate between valid and invalid?

But doing nothing means that the SDK can emit OTLP data that violates the monotonicity description. An approach I think is reasonable (and others may have agreed - can't quite remember) is to do nothing here. By doing nothing we take the stance that the SDK is going to do its best to validate recordings and produce valid data. But occasional edge cases like this one may prevent that. It's up to backends to handle this in an appropriate way.

Reply:
> I don't believe we have any precedent for reseting the start timestamp yet - I'm interested in ideas on how this would work.

We have some preliminary discussion from > 1 year ago, when the data-model SiG was meeting regularly.  Specifically, we do discuss [reset here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#resets-and-gaps).   We decided, in the end, to attempt to model resets using purely timestamps.  This is purely our datamodel.  There are practical implications to this.


If you're interested in the design space and want to look into solutions, I suggest diving into [this issue](https://github.com/prometheus/prometheus/issues/6669) which dives deep into practical concerns with async counters and expectations of reset in SUMs *plus* the major issue happens from OpenMetrics/Prometheus forcing positive-only histograms.

Additionally, take look at why cumulative values dropping is ok in practice:  the [resets](https://prometheus.io/docs/prometheus/latest/querying/functions/#resets) function in PromQL.

There's a great discussion to be had over whether the SDK should do reset-like funcitonality to ensure the PromQL `reset` method and our start/stop timestamps (specified in our data model) align.   AFAIK, not many backends use (only) timestamps for reset detection.  But most will use value drops (particularly anything touting promql compatibility).

> We talked in this in the 12/16 6pm PST SIG. The SDK is in a tough spot with async counters because there's no natural way to validate that values are monotonically increasing. This approach checks in the aggregator, which means that the validation rules change based on how you're aggregating, which isn't right - an async counter doesn't all the sudden get to break monotonicity because the aggregation is last value. And to your point, roll-overs are valid - how do we differentiate between valid and invalid?

My point is I don't think we can.   We should assume good intent from the user (and document expectations) in these situations.  If you come up with something that won't break prometheus expectations, then I'd be all for something better.

TL;DR; I still think doing nothing here is better than preventing recordings.  If you'd like to log (and still record) that's fine as well.

Reply:
> TL;DR; I still think doing nothing here is better than preventing recordings. If you'd like to log (and still record) that's fine as well.

I've rolled it back. The log message isn't worth the complexity to implement.